docat/docat/utils.py

```diff
@@ -90,3 +97,230 @@ def calculate_token(password, salt):
+
+    for version in details.versions:
+        insert_version_into_version_index(index_db_path, project, version.name, version.tags)
+
+
+def get_html_content_as_str(file_path: Path):
+    """
+    Returns the content of a html file as a string.
+    """
+
+    def html_tag_visible(element):
+        if element.parent.name in ["style", "script", "head", "title", "meta", "[document]"] or isinstance(element, Comment):
+            return False
+
+        return True
+
+    with open(file_path, "r") as f:
+        file_content = f.read()
+        soup = BeautifulSoup(file_content, "html.parser")
+        text_content = filter(html_tag_visible, soup.findAll(string=True))
+        content = " ".join(t.strip() for t in text_content).lower()
+        return content
+
+
+def insert_file_index_into_db(index_db_path: Path, project: str, version: str, file_path: str, content: str):
+    """
+    Inserts a file index into the index.json.
+    """
+    index_db = TinyDB(index_db_path)
+    files_table = index_db.table("files")
+
+    files_table.insert({"path": file_path, "content": content, "project": project, "version": version})
+
+    index_db.close()
+
+
+def remove_file_index_from_db(index_db_path: Path, project: str, version: str):
+    """
+    Removes the file index for the given project version
+    """
+
+    index_db = TinyDB(index_db_path)
+    files_table = index_db.table("files")
+
+    File = Query()
+    files_table.remove(File.project == project and File.version == version)
+
+    index_db.close()
+
+
+def insert_version_into_version_index(index_db_path: Path, project: str, version: str, tags: list[str]):
+    """
+    Inserts a project index into the index db.
+    """
+    index_db = TinyDB(index_db_path)
+    projects_table = index_db.table("projects")
+    Project = Query()
+    found_projects = projects_table.search(Project.name == project)
+
+    if not found_projects:
+        # create
+        projects_table.insert({"name": project, "versions": [{"name": version, "tags": tags}]})
+        index_db.close()
+        return
+
+    existing_versions = found_projects[0].get("versions")
+
+    if not existing_versions:
+        return  # should not happen
+
+    if version in (v.get("name") for v in existing_versions):
+        # version already exists, remove so we can add it again, updating the tags
+        existing_versions = list((v for v in existing_versions if v.get("name") != version))
+
+    existing_versions.append({"name": version, "tags": tags})
+    projects_table.update({"versions": existing_versions}, Project.name == project)
+    index_db.close()
+
+
+def remove_version_from_version_index(index_db_path: Path, project: str, version: str):
+    """
+    Removes a version from the project index in the index db.
+    """
+    index_db = TinyDB(index_db_path)
+    projects_table = index_db.table("projects")
+
+    Project = Query()
+    found_project = projects_table.search(Project.name == project)[0]
```

Comment:
This will raise an index out of range error when `[]` is returned, check `if not found_project` needs to be done before the index is accessed